src/TypeChecker/Inference.hs

```diff
@@ -0,0 +1,521 @@
+
+constructorsToEnv ((TypeCase (ConstructorName name) types _ ):tail) 
+                  typeName 
+                  params 
+                  env =
+  case params of
+    [] -> 
+      let env' = extendEnv env (TypeVar name, schemeNoParams)
+      in constructorsToEnv tail typeName params env'
+    xs -> 
+      let env' = extendEnv env (TypeVar name, schemeParam)                               
+      in constructorsToEnv tail typeName params env'
+  where 
+    schemeNoParams = Scheme [] (foldr FunctionType 
+                                      (NonPrimType typeName) 
+                                      types)
+    schemeParam = Scheme params 
+                         (foldr FunctionType 
+                                (ParamType typeName (map TypeVariable params)) types)
+      
+
+typecheckAlias :: TAlias -> Environment -> Either String Environment
+typecheckAlias (TAlias (TypeName name) params kind alType _ ) env =
+  -- TODO: check if params are in type / check type correctness?
+  --        or is it done in parsing? 
+  -- TODO: rework
+  Right $ extendEnv env ((TypeVar name), Scheme params alType)
+
+
+typecheckTopLevelLet :: Definition a -> Environment -> Either String Environment
+typecheckTopLevelLet def env = case letPattern def of
+  (VarPattern (Identifier id) pos _ ) -> 
+    let res = inferLetDefExpr (letExpr def) env
+    in case res of
+        Left err -> Left err
+        Right t  -> 
+          let resUni = unifyWithTypeSig t (letTypeSig def)
+              vars = (Set.toList (freeTypeVar t))
+          in case resUni of
+              Left err -> Left err
+              Right _ -> Right (extendEnv env (TypeVar id, Scheme vars t))
+  p@(TuplePattern _ pos _ ) -> 
+    let resPattern = inferLetDefComplexPattern p def env
+    in case resPattern of
+        Left err -> Left err
+        Right (env', exprType) -> 
+          let resUni = unifyWithTypeSig exprType (letTypeSig def)
+          in case resUni of
+              Left err -> Left err
+              Right _ -> Right env' 
+  p@(ConstructorPattern _ _ pos _ ) -> 
+    let resPattern = inferLetDefComplexPattern p def env
+    in case resPattern of
+        Left err -> Left err
+        Right (env', exprType) ->
+          let resUni = unifyWithTypeSig exprType (letTypeSig def)
+          in case resUni of
+              Left err -> Left err
+              Right _ ->  Right env' 
+  (NamedPattern (Identifier id) pattern pos _ ) ->
+    let resPattern = inferLetDefComplexPattern pattern def env
+    in case resPattern of
+      Left err -> Left err
+      Right (env', exprType) -> 
+        let resUni = unifyWithTypeSig exprType (letTypeSig def)
+            vars = (Set.toList (freeTypeVar exprType))
+          in case resUni of
+            Left err -> Left err
+            Right _ -> 
+              Right $ extendEnv env' (TypeVar id, Scheme vars exprType)
+
+  _ -> Left $ "Error: Illegal operation"
```

Comment:
It should be "Illegal pattern type in let definition" or something like that imho